38/LOGOS-CONSENSUS-GLACIER

easye opened the pull request:
c.f. <https://github.com/vacp2p/rfc/issues/509>

Ongoing pull request towards an initial Logos Consensus RFC

easye committed:
38/LOGOS-CONSENSUS RFC initial TOC

c.f. <https://github.com/vacp2p/rfc/issues/509>

kaiserd commented:
Thanks for starting this PR.
Imo, the focus should be on aspects that are important for implementors.
So, the more background related sections could be left as placeholders / filled with a rough sketch for now.

kaiserd commented:
I'd suggest adding a section for  `Wire Format Specification / Syntax` after  `Exposition of Snow-family Consensus Models`
( In #488, I generically described what could/should be in that section.)

Also adding a "security" section like `Security/Privacy Considerations` whould be helpful.
The contents can be a rough sketch for the raw version of the RFC.

You can also use the following copyright section:

```
￼# Copyright
￼
￼Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,21 @@
+slug: 38
+title: 38/LOGOS-CONSENSUS
+name: Initial Logos Consensus Description
+status: incomplete-pre-draft
```
```suggestion
status: raw
```

kaiserd replied:
I'd stick to making the status one of `raw`, `draft`, `stable`.
In this case `raw` is what this PR works towards.

easye committed:
Change status; add copyright waiver

Addresses
<https://github.com/vacp2p/rfc/pull/512#issuecomment-1172356790>, <https://github.com/vacp2p/rfc/pull/512#discussion_r911972504>.

easye committed:
SQUASHME:  Start of transcription of written notes.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,59 @@
+# Consensus
+# Interoperability 
+## Wire Formats Specification / Syntax
+At this point, we do not commit to a syntax but rather to a semantics
```
I'd open a new section (above this one) for the semantics. 

Somboday who just wants to implement, and already has an understanding of how things work, can jump to the sytax section for reference and start implementing.

easye replied:
Hmm.  The subtlety is there isn't a wire protocol defined yet, just a semantics of what meanings nodes exchange for consensus rounds.  I was thinking about deriving an interoperable JSON syntax from an abstract semantics, hence this approach.  

I understand and accept your emphasis on ease of implementation here, so will reconsider my approach when I finish getting my PR draft for today ready to push.

kaiserd replied:
Wdyt @oskarth @corpetty? Is there any raw material a wire format spec (e.g. protobuf) could be based on?
In case we don't have this, i'd suggest starting with the semantics section and leave the background/intro/motivation for last.
Imo, starting with the semantics section is perfectly fine, too.
Mainly, my point was to put the semantics in a separate section like so:

```

## Theory / Semantics

<If there is no sufficient material to start with a raw wire spec, start here>

## Wire Formats Specification / Syntax

At this point, we do not commit to a syntax but rather to a semantics [...]

```

easye replied:
The Theory section has been moved to appendices, being replaced with a complete pseudo-code transcription of Glacier.

For wire formats, users are advised to construct a suitable Waku payload with appropiate metada probaly via JSON, preferrably via JSON-LD.

oskarth commented:
Where's the content? I see nothing that relates to actual consensus mechanism. It shouldn't take that long to get basic semantics done considering a lot has been written already

D4nte commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,59 @@
+## In Media Res
+
+We begin--as all worthwhile journeys begin--in the middle of things,
+desiring to communciate a whole without knowing the exact shape of the
```
communicate*

The use of semantic line breaks https://sembr.org/ is recommended https://rfc.vac.dev/spec/1/#conventions

easye replied:
> The use of semantic line breaks https://sembr.org/ is recommended https://rfc.vac.dev/spec/1/#conventions

Choosing to not implement this suggestion for the moment, as manual paragraph filling is part of my re-writing process. 

  - [ ] But I'll definitely take this as a TODO for exiting the RAW specification state to not confuse upstream consumers.

easye replied:
It doesn't seem that the inclusion of line breaks is affecting downstream Markdown consumer's notion of paragraphs.  

But will double check before we enter ~~draft~~ raw.

easye committed:
Commit before attempting to merge with master

easye committed:
Describe where we are going in the abstract

easye committed:
Copy and paste Glacier description

Use of inline LaTeX can be rendered via

    pandoc -f commonmark_x README.md -o README.pdf

easye committed:
Add revised requirements from Moh; start adjusting math markup

easye committed:
Start re-working requirements and Glacier description

easye committed:
First plausible Glacier description

easye committed:
Simplify scope

Strip out all background other than explaining where Glacier would fit
into a consensus mechanism useful for voting on transitions for a
shared state machine.

easye committed:
Correct pseudo code for misname variable

easye committed:
Further re-working

easye committed:
Hopefully worked through all the LaTex markdown issues

easye committed:
Maybe all the pseudo-code for Glacier is legit

easye committed:
Maybe complete pseudo-code

TODO:  threshold for confidence

easye committed:
Fix runaway network multiplier.

TODO: check current constants with simulation.  From looking at
Alexander's graphs, I think choosing a low initial k of seven--the
first prime under a nibble--with an agressive expansion by factors of
two makes a good compromise.  Glacier converges surprisingly fast, and
since it reconstructs the neighbors queried in each round, it is
suitably probabilistic.

easye commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,480 @@
+       node == UNDECIDED
+    THEN 
+       k   ;; number of nodes to uniformly randomly query in next round
+         <-- $k * \sup{ k_multiplier, max_k_multiplier}$
```
This is bogus.

easye replied:
should be 

    k *-- $sup{k * k_multiplier, k_multiplier * max_k_multiplier}

easye replied:
I don't understand why I keep missing this.  *sigh*

easye commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,480 @@
+---
```
This a comment

oskarth commented:
What is needed before this is ready for review / mergeable? What about being draft (does this even make sense given current state?) Does it reflect and capture all relevant current knowledge in e.g. Notion?

easye commented:
> What is needed before this is ready for review / mergeable? What about being draft (does this even make sense given current state?) Does it reflect and capture all relevant current knowledge in e.g. Notion?

- [ ] Camera-ready edit of text
- [x] Confirm how markdown renders locally (using Pandoc 2.18 current; Github preview is not ready)
- [x] Alvaro's contributor status clear
- [x] pseudo-code description of Glacier accurate 
- [ ] review by end of August from at least two Logos contributors
   - [x] Increase Logos Github familiarity for those who need it

easye commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,480 @@
+      <-- 2
+
+    ;; maximal threshold multiplier, i.e. we will never exceed 
+    ;; questioning k_multiplier * max_k_multiplier peers
```
err 

    k_multiplier ^ max_k_multiplier

AMozeika commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,480 @@
+algorithm is severely constrained in the absence of timeouts for a
+round to proceed.
+
+### Setup
```
I guess to name this section "Parameters" would more accurate. The total number of nodes N (or list of nodes)  is also needed for this algorithm to operate? Also I think that a parameter which tells the algorithm when to stop is missing  in this section.

easye replied:
Changed to "Setup Parameters".  Distinguished between constants and variables in their use in the rounds.

AMozeika commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,480 @@
+values:
+
+ 
+    new_votes 
```
Is this just $k$ when all nodes are available?

easye replied:
Not all votes are necessarily collected in a round (time outs).

AMozeika replied:
Ok

AMozeika commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,480 @@
+
+ 
+    new_votes 
+      <-- |vote replies received in this round|
```
"round" is equivalent to "query"?

easye replied:
Yep.  Tried to make that more clear.

AMozeika commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,480 @@
+    ;; total number of votes
+    total_votes 
+      <-- 0    
+    ;; total number of positive nodes
```
"total number of positive votes"?

AMozeika commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,480 @@
+      += positive_votes
+    confidence 
+      <-- total_votes / (total_votes + look_ahead) 
+    evidence_accumulated 
```
Perhaps  for  "evidence" we could use the same convention as for "votes", i.e. to use "total_evidence" instead of "evidence_accumulated"  and "new_evidence" for "evidence_round"?

easye commented:
> Where's the content? I see nothing that relates to actual consensus mechanism. 

A base pseudo-code exposition of Glacier now exists.  It needs editing, but is largely correct.  Main problem is describing termination  accurately.

easye committed:
Addressed Alexander's review

Pseudo-code mostly correct.

TODO tighten the algorithm termination description (currently bogus)

easye commented:
> I'd suggest adding a section for `Wire Format Specification / Syntax` after `Exposition of Snow-family Consensus Models` 

Resolved.
> `Security/Privacy Considerations` whould be helpful. 

See the "Sovereignty Considerations" section. Thanks!

> ￼# Copyright

Added.

easye committed:
Get rid of float constants in psudeo-code

easye commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,507 @@
+in the query round through the following algorithm:
+
+    total_votes 
+      += new_votes
```
TODO Change to from `+=` to `+==` normalizing to three characters like all other "line join" operators.

easye commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,507 @@
+The node updates its local opinion on the consensus proposal:
+
+    IF
+      $evidence$ > $alpha$ 
```
Remove LaTex prefixes here.  Include only in appendices until we debug the presentation run.

easye commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,507 @@
+to use Waku messages to include their own metadata in serializations
+as needed.  Such a message contains minimally
+
+    round ;; the local number of rounds participating 
```
Write out as N3 Bags for semantics.

oskarth commented:
I think it would it make sense to have a brief section on how this compares with PoW/Nakamoto Consensus and PoS/LMD-GHOST (https://arxiv.org/pdf/2003.03052.pdf)? Similarities and differences, briefly. It should explain to the reader at a glance why a new consensus mechanism is even needed.

(This short summary can then be elaborated on in its own research post, but this is out of scope for this spec)

oskarth commented:
Just looking briefly at https://www.notion.so/Consensus-cca3f5b9bed341219d1cfd027bea897c the first thing I see is a lot of helpful requirements regarding fast finality etc. This also might disqualify other mechanisms. I can't find the same in this spec, is there a reason why?

oskarth commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,507 @@
+implementation of the base binary decision mechanism.  We outline a
+taxonomy of Byzantine adversaries that seek to thwart this
+computation's correct honesty.
+
```
Requirements would fit nicely here

oskarth replied:
Ditto something like "prior work" or "motivation" (comparing e.g. BTC/ETH consensus briefly)

easye replied:
I'll make a stab at further explaining the prior work of leaderless BFT binary consensus.

easye commented:
> Just looking briefly at https://www.notion.so/Consensus-cca3f5b9bed341219d1cfd027bea897c the first thing I see is a lot of helpful requirements regarding fast finality etc. This also might disqualify other mechanisms. I can't find the same in this spec, is there a reason why?

There does not exist Logos "consensus" on these requirements.  There are the original requirements that Alvaro specified, Mo's revised requirements (in progress), and requirements for Status Communities which would take some, eh, consensus to work through to a single set.  The abstraction of requirements is incomplete, in that it doesn't provide tracibility (what is the trace of the chain from which we derive), and doesn't do basic functional/non-functional etc. separation.  

I will make a first pass at harmonization to see if Logos would provisionally agree.

easye committed:
Stab at message semantics in N3

fryorcraken commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,534 @@
+
+We propose to transition our use of resource intensive Nakomoto
+consensus mechanisms to the class of more efficient, secure,
+leaderless ones rooted in decentralization.  We sketch a simple
```
semantic break line is encouraged: https://rfc.vac.dev/spec/1/#conventions

easye commented:
> I think it would it make sense to have a brief section on how this compares with PoW/Nakamoto Consensus and PoS/LMD-GHOST (https://arxiv.org/pdf/2003.03052.pdf)? Similarities and differences, briefly. It should explain to the reader at a glance why a new consensus mechanism is even needed.

The essential advantage of Glacier over PoW/Nakamoto consensus is efficiency; the advantage over PoS/LMD-GHOST is the leaderless nature.  Both of these qualities are currently mentioned in the first sentence of the abstract.  But Glacier in and of itself is not a "blockchain" consensus mechanism.  It is better thought of as a BFT primitive for a single decision that would contribute to a larger "blockchain" set of such decisions.

> It should explain to the reader at a glance why a new consensus mechanism is even needed.

The more the merrier?  Seriously, without a whole slew of political justifications, this moves quickly away from being an RFC about an implementation, to a manifesto for Logos.  Original versions of this pull request tried to describe these justifications, rather incompletely no matter how much text I threw at it (and then took away).   On review, @kaiserd suggested that I  concentrate on the actual algorithm, esp. wrt. being able to implement it.  Glacier just provides yes/no votes on a given proposal.  How to turn that mechanism into a "blockchain" consensus is rather involved, with Logos not having an organic expression of how to do it.  All this is complicated by the unfinished analysis of the Byzantine (or even simple fault) tolerance of Glacier security which could use a stand-alone description of Glacier to proceed.

But I will see what I can add to the existing sections to strengthen the motivation of a reader.  Maybe an Appendix with some of the Logos manifesto stuff?

easye committed:
Morning edits

easye committed:
Rename to 38/LOGOS-CONSENSUS-GLACIER

oskarth commented:
I think just hammering home the leader-less aspect and efficiency aspect of it and contrasting this with PoW/Nakamoto and PoS/LMD-GHOST would be useful. Can be a short paragraph, just so someone can have "aha" moment. Even though strictly speaking it is mentioned already.

Agree re not making it political at this stage, more about explaining property briefly.

danielSanchezQ commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,556 @@
+    
+    ;; possibly increase number nodes to uniformly randomly query in next round
+    WHEN
+       opinion == UNDECIDED
```
This may be confusing, we should not mix `opinion` and `decision`:
* Opinion: yes, no, none
* Decision: Variable(`undecided`)/invariable(`decided`)  node state

easye replied:
Changed UNDECIDED opinion to NONE

danielSanchezQ commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,556 @@
+    OR 
+      round > max_rounds
+    THEN
+      finalized <-- T
```
This finalized flag could be the `decision` itself.

easye replied:
Hmm, I haven't specified the Termination Conditions.  I was in the middle of refactoring it out in a section to describe modeling/running in async/sync modes.  I'll TODO RSN.

danielSanchezQ commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,556 @@
+If the view of other nodes is incomplete, then the sum of the optional
+weighting won't be a probability distribution normalized to 1.
+
+The current algorithm doesn't describe how the initial opinions are formed.
```
By default an honest node would vote `yes` for anything that is not in conflict with previous voted things (and finalised ones). Otherwise is up the heuristics of resolving conflicts. But, for a happy path, every honest nodes always votes on `yes`.

danielSanchezQ commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,556 @@
+delay, and drop arbitrary messages in the gossip layer, and utilize
+this to corrupt consensus away from honest decisions to ones favorable
+to itself.  It may, of course, choose to participate in an honest
+manner until defecting is most advantageous.
```
One thing missing here is that we consider the Omniscient adversary to be fully aware at all times of the network state. Meaning that he can always chose to vote in the most efficient way to block the network from reaching finality.

easye replied:
TODO Need to start dividing the Omniscient Adversary along complexity classes and coordination power.

danielSanchezQ reviewed (commented):
Left some comments, overall looks nice 😄

easye committed:
Iterating through how Github LaTex markdown works

easye committed:
Iterating LaTeX

easye committed:
math

easye committed:
math

easye committed:
math

easye committed:
Do arrays work?

easye committed:
hugo:  add MathJax for all RFCs via template

TODO: examine if we really want this, as it causes an additional
JavaScript load for pages that aren't using LaTeX

easye committed:
math that renders in Hugo via MathJax

Does this work for Github?

easye committed:
math

easye committed:
Reasonble version of Appendix A?

easye committed:
Most math rendering:  exception subscript of multi chars

easye committed:
Attempt to get \_ working for _

easye committed:
mathjax:  skip HTML tags for which we don't allow LaTeX

easye committed:
Addressed @danielSanchezQ comments

Rename UNDECIDED opinion as NONE.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,563 @@
+that seek to thwart this computation's correct honesty.
+
+# Consensus Model
+
```
Imo, an introductory sentence would help making the following paragraph easier to digest.
Something like: "The consensus model specified in this document is based on the fact that given an underlying [...]"

corpetty replied:
agreed, and this justification can come from Alvaro's original document in Notion. It specifically describes what the algorithm aims to improve over current Snow protocols.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,563 @@
+---
+
+# Abstract
+
```
As this is an RFC, I'd start with (copied from the notion document; changed "describes -> specifies"):

------

This document specifies the Glacier Binary Byzantine Agreement algorithm, and its extension to a DAG structure in order to achieve a leaderless byzantine-resistant algorithm for consensus, of application in the context of permission-less blockchains.

------

It should be very clear, what this document actually specifies.

easye replied:
The "and its extension to a DAG structure" part should be elided, as we are not even attempting to specify how to use Glacier.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,563 @@
+contract" artifacts by implementing a given model of transaction.
+
+# Glacier 
+
```
I'd mention that `Glacier` is the algorithm that is specified in this document.

In the future, it might make sense to actually introduce/propose the algorithm in a scientic paper,
and then write here: "... This document speciefies the (integration / usage of) Glacier, proposed in [?], in the context of logos..."

kaiserd commented:
>  > It should explain to the reader at a glance why a new consensus mechanism is even needed.

> The more the merrier? Seriously, without a whole slew of political justifications, this moves quickly away from being an RFC about an implementation, to a manifesto for Logos. Original versions of this pull request tried to describe these justifications, rather incompletely no matter how much text I threw at it (and then took away). On review, @kaiserd suggested that I concentrate on the actual algorithm, esp. wrt. being able to implement it. Glacier just provides yes/no votes on a given proposal. How to turn that mechanism into a "blockchain" consensus is rather involved, with Logos not having an organic expression of how to do it. All this is complicated by the unfinished analysis of the Byzantine (or even simple fault) tolerance of Glacier security which could use a stand-alone description of Glacier to proceed.

I agree with both.
Imo the best apporach would be proposing Glacier in a scientific paper, which explains Glacier's strength and why it is needed.
The RFC can then cite this paper and stay neutral.
In a raw RFC, it would be find to include such reasoning in the RFC.

> I think just hammering home the leader-less aspect and efficiency aspect of it and contrasting this with PoW/Nakamoto and PoS/LMD-GHOST would be useful. Can be a short paragraph, just so someone can have "aha" moment. Even though strictly speaking it is mentioned already.

I agree, such a paragraph would be helpful in the "Consensus" section, which is a "background" section and not part of the actual spec. It should be a non-opinionated, factual explanation.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,563 @@
+The node then participates in a number of query rounds in which it
+solicits other node's opinion in query rounds.  Given a set of $n$
+leaderless computational nodes, a gossip-based protocol is presumed to
+exist which which allows members to discover, join, and leave a weakly
```
```suggestion
exist which allows members to discover, join, and leave a weakly
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,563 @@
+solicits other node's opinion in query rounds.  Given a set of $n$
+leaderless computational nodes, a gossip-based protocol is presumed to
+exist which which allows members to discover, join, and leave a weakly
+transitory maximally connected graph.  Joining this graph allows each
```
Does "weakly transitory" mean the algorithm assumes a low churn rate?

easye replied:
I was more envisioning considering a dynamically varying churn, at different timescales.  Nodes enter and leave as the protocol rounds progress.

kaiserd replied:
I'd make that explicit.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,563 @@
+leaderless computational nodes, a gossip-based protocol is presumed to
+exist which which allows members to discover, join, and leave a weakly
+transitory maximally connected graph.  Joining this graph allows each
+node to view a possibly incomplete node membership list of all other
```
Assuming the graph is maximally connected, why is the node list "possibly incomplete".
Is that due to chrun? Byzantine faults (due to connectivity problems / deliberate attacks)? I'd explicitly state that.

easye replied:
> why is the node list "possibly incomplete".
>Is that due to chrun? [sic]

Yes, faults both Byzantine and otherwise (hardware/software/network).

kaiserd replied:
I'd explicitly state this as the reason why we might not see the whole memeber list even though we have a full graph.

easye commented on resources/_gen/assets/scss/book.scss_50fc8c04e12a2f59027287995557ceff.content:
```diff
@@ -1 +1 @@
-:root{--gray-100: #f8f9fa;--gray-200: #e9ecef;--gray-500: #adb5bd;--color-link: #0055bb;--color-visited-link: #8440f1;--body-background: white;--body-font-color: black;--icon-filter: none;--hint-color-info: #6bf;--hint-color-warning: #fd6;--hint-color-danger: #f66}/*!normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css*/html{line-height:1.15;-webkit-text-size-adjust:100%}body{mar ...
```
This change needs to be elided from the final pull request.

oskarth replied:
@corpetty this should be unstaged, ditto .hugo_build.lcok file

easye commented on layouts/docs/single.html:
```diff
@@ -1,4 +1,26 @@
 {{ define "main" }}
+
```
This includes MathJax in all RFCs.  Supposedly the JavaScript hooks run on load does not parse the DOM nodes whose names match any of the values of `skipTags`, but this has not been verified.

This `SCRIPT` load relies on a CDN for which we should either a) host ourselves, or b) declare a checksum for the contents.

easye committed:
Update content/docs/rfcs/38/README.md

via @kaiserd

Co-authored-by: kaiserd <git@kais3r.de>

easye committed:
Clarify abstract.

_0: :Addresses <https://github.com/vacp2p/rfc/pull/512#discussion_r949149399>.

easye committed:
Reorder opinion as NO, NONE, and YES

easye committed:
squash to previous commit

easye committed:
Remove LaTeX in favor of Markdown

easye committed:
Tit-for-tat

easye committed:
Mention Logos

easye committed:
Experimentally tighten the pseudo-code

easye committed:
Close markdown

TODO:  find a lightweight linter

easye committed:
Whack a 'TeX

easye committed:
Another round of whack-a-Github Markdown

easye committed:
/Ibid/

easye committed:
Possible point to bisect for triage of LaTeX woes

easye committed:
Pay attention to unescaped use of left square brackets

easye committed:
Another whack

easye committed:
Correct N3; reference W3C 2011 standard

easye committed:
Note JSON-LD while whackin' the 'Tex

easye commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,650 @@
+
+Although we have proposed a normative description of the
+implementation of the underlying binary consensus algorithm (Glacier),
+we believe we have analyzed its adversarial performance in a manner
```
"we believe we have prepared for analysis"

easye commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,650 @@
+1. [On BFT Consensus Evolution: From Monolithic to
+   DAG](https://dahliamalkhi.github.io/posts/2022/06/dag-bft/)
+
+2. [snow-ipfs](https://ipfs.io/ipfs/QmUy4jh5mGNZvLkjies1RWM4YuvJh5o2FYopNPVYwrRVGV)
```
Add ipfs URI.

easye commented on content/menu/index.md:
```diff
@@ -11,6 +11,7 @@ bookMenuLevels: 1
   - [32/RLN-SPEC]({{< relref "/docs/rfcs/32/README.md" >}})
   - [35/WAKU2-NOISE]({{< relref "/docs/rfcs/35/README.md" >}})
   - [37/WAKU2-NOISE-SESSIONS]({{< relref "/docs/rfcs/37/README.md" >}})
+  - [38/LOGOS-CONSENSUS]({{< relref "/docs/rfcs/38/README.md" >}})
```
38/LOGOS-CONSENSUS-GLACIER

easye committed:
Reference how to measure local clock drift

Add better normative N3 reference.

easye committed:
Language edits

easye committed:
More linting

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+models of the finalization of shared data structure, Logos starts the
+construction of a "value-free" Network State infrastucture.
+Subsequently, we will connect the implementation of Glacier to an the
+appopiately secure execution of a explicitly desired security model.
```
```suggestion
appropriately secure execution of a explicitly desired security model.
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+documenting Glacier, a component for shared consenus composable with
+models of the finalization of shared data structure, Logos starts the
+construction of a "value-free" Network State infrastucture.
+Subsequently, we will connect the implementation of Glacier to an the
```
```suggestion
Subsequently, we will connect the implementation of Glacier to an
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+Nakomoto consensus mechanisms with those which belong to the class of
+more efficient ones rooted in leaderless decentralization.  By
+documenting Glacier, a component for shared consenus composable with
+models of the finalization of shared data structure, Logos starts the
```
```suggestion
models of the finalization of shared data structures, Logos starts the
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+construction of a "value-free" Network State infrastucture.
+Subsequently, we will connect the implementation of Glacier to an the
+appopiately secure execution of a explicitly desired security model.
+As a member of the Snow family, Glacier provides an tractably
```
```suggestion
As a member of the Snow family, Glacier provides a tractably
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+computable probabilistic measure of safety of the
+finalization of states of shared data structures.
+
+We start with simple two-level composition model for the practical
```
```suggestion
We start with a simple two-level composition model for the practical
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+We start with simple two-level composition model for the practical
+execution of such a fairly distributed consensus mechanism, in which
+an underlying leaderless binary decision mechanism votes on the
+finalization of nodes distributed, directed, acyclic graph.  Here, we
```
[...] votes on the finalization of nodes distributed, directed, acyclic graph. 

I am not sure how attributes and nouns relate here.
Maybe "nodes distributed in a directed, acyclic graph"?

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+finalization of nodes distributed, directed, acyclic graph.  Here, we
+present a pseudo-code specification the Glacier algorithm which
+provides a Byzantine fault tolerant implementation of the base binary
+decision mechanism.  We start to outline a taxonomy of Byzantine adversaries
```
```suggestion
decision mechanism. We start to outline a taxonomy of Byzantine adversaries
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+subsequent results need to be annotated.
+
+The algorithim we specify here, Glacier, may be extended to a DAG
+structure in order to achieve a leaderless byzantine fault tolerancr.
```
```suggestion
structure in order to achieve a leaderless byzantine fault tolerance.
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+name: Logos Glacier Consensus Protocol
+status: raw
+category: informative
+tags: logos/consensus,implementation/rust, implementation/python, implementation/common-lisp
```
```suggestion
tags: logos/consensus
```

I'd remove implementation related tags.
The tags were introduced in #368 

The idea was to have a specified list of tags. We could have an RFC listing tags...
Alternatively, we could introduce tags in a more dynamic less structured manner.
I'll start a discussion on that (meeting and/or discord). Feel free to comment here, too :).

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+title: 38/LOGOS-CONSENSUS-GLACIER
+name: Logos Glacier Consensus Protocol
+status: raw
+category: informative
```
If this document actually specifies Glacier, it should be `Standards Track`.
An alternative would be introducing two documents. The `Informational` could cover things that alternatively could be part of a scientifc paper ([see my comment](https://github.com/vacp2p/rfc/pull/512#issuecomment-1219518824)), while the `Standards Track` contains the actual specification.

easye committed:
Update content/docs/rfcs/38/README.md

Co-authored-by: kaiserd <git@kais3r.de>

easye committed:
Update content/docs/rfcs/38/README.md

Co-authored-by: kaiserd <git@kais3r.de>

easye committed:
Update content/docs/rfcs/38/README.md

Co-authored-by: kaiserd <git@kais3r.de>

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+execution of such a fairly distributed consensus mechanism, in which
+an underlying leaderless binary decision mechanism votes on the
+finalization of nodes distributed, directed, acyclic graph.  Here, we
+present a pseudo-code specification the Glacier algorithm which
```
```suggestion
present a pseudo-code specification of the Glacier algorithm which
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+;;; constant look ahead for number of rounds we expect to finalize a
+;;; decision.  Could be set dependent on number of nodes 
+;;; visible in the current gossip graph.
+look_ahead 
```
Does this convey any information beyond the number of expected rounds to finalize?
If not, why not call it something like `num_expected_rounds`?

easye replied:
Strictly speaking, `look_ahead` is directly porportional to number of rounds, and is not exactly the magnitude of rounds.  Renaming makes sense.  The term was Alavaro's original.

kaiserd replied:
The term was confusing to me in that context because you can't actually look ahead like you can in regexes.

In which way is it directly porportional to the number of rounds?

easye replied:
`look_ahead` is used to compute a `confidence` value:
```
    confidence 
      <-- total_votes / (total_votes + look_ahead) 
```
which is directly proportional `evidence` and `alpha` parameters used in the DECISION phase:
```
    evidence 
      <-- new_evidence * ( 1 - confidence ) + total_evidence * confidence 
    alpha 
      <-- doubt * ( 1 - confidence ) + certainty * confidence 
```

Thus, in rounds where `total_votes` << `look_ahead`, Glacier does not put much faith in what it learns from the queries.

kaiserd replied:
Yes, this is how `look_ahead` is used. But, in which way is the value of `look_ahead` itself proportional to the number of rounds?
Is there another factor besides being the expected number of rounds?

If it is just the expected number of rounds, I'd rename that variable to `num_expected_rounds`.
Imo, the following statement would be easier to grasp, too, if `look_ahead` would be replaced with `num_expected_rounds`:

> Thus, in rounds where total_votes << look_ahead, Glacier does not put much faith in what it learns from the queries.

Any other opinions here? Is `look_ahead` established terminology for this?

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+    IF
+      evidence > alpha
+    THEN ;; The node adopts the opinion YES on the proposal
```
Imo, the comments, e.g. `;; The node adopts the opinion YES on the proposal`, make this part harder to read.
The pseudo code in itself is obvious. I'd remove these comments.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+      evidence < 1 - alpha
+    THEN ;; The node adopts the opinion NO on the proposal 
+      opinion <-- NO
+       
```
I'd make the else case explicit here and add

```
      ELSE 
￼           opinion <-- NONE
```

easye replied:
Hmm.  Glacier never sets its `opinion` to `NONE` in this round, this is only possibly done in the initialization step.  If after checking the `confidence` and `alpha` relationship and if the opinion is `NONE`, Glacier widens its search by increasing the number of nodes it queries in the next round.

Perhaps I need to approach this section differently to make that more clear?

kaiserd replied:
In this case, I'd just add a clause saying that if none of the conditional branches are hit, the opinion is/stays NONE.

When I read this

> If the opinion of the node is `NONE` after evaluating the relation between `evidence` and `alpha`

I assumed, option is NONE in the default case, but it would be easier to grasp if that would be explicit.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+;;; maximal threshold multiplier, i.e. we will never exceed 
+;;; questioning k_initial * k_multiplier ^ max_k_multiplier peers
+max_k_multiplier 
```
This naming is confusing (at least for me it was when I read the formula using it below)

The actual max multiplier is ` k_multiplier ^ max_k_multiplier`

I'd either

* call this variable  `max_k_multiplier_power`, or
* set `max_k_multiplier = k_multiplier ^ 4`

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+is considered: if `confidence` exceeds a threshold derived from the
+network size and directly related to the total votes received, an
+honest node marks the decision as final, and always returns this
+opinion is response to further queries from other nodes on the
```
```suggestion
opinion as a response to further queries from other nodes on the
```

kaiserd replied:
(Just marking this as still in there during my current review.)

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+      QUERY LOOP TERMINATES
+    ELSE 
+      round +== 1
+      GOTO QUERY 
```
Where is QUERY? If it is not explicitly defined somewhere, I'd rather say `QUERY LOOP CONTINUES`,
matching the previous phrasing of `QUERY LOOP TERMINATES`

kaiserd replied:
(Just saw, the section is called query. I'd still go with the suggestion in my previous comment as `QUERY` is not defined as label in the code.)

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+The algorithm terminates under the following execution model considerations:
+
+A local round of Glacier terminates in one of three conditions:
```
I'd merge these sentences.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+
+1.  No queries are received for any newly initiated round for temporal
+    periods observed via a locally computed passage of time.
```
Are there any constraints on the accuracy of calculating said period?

kaiserd replied:
(I'd assume, if a node happened to have "a clock whose seconds go faster", this node would be considered byzantine.)

easye replied:
There is a section on the notion of "local clock" below which I have referenced.  

For Glacier nodes who are accelerating at appreciable relativistic velocities, they will definitely have trouble participating in consensus as they essentially never initiate queries.  For nodes that have "a clock whose seconds go faster", I think the basics of the algorithm hold ok.  Need to think a bit if there is a simpler way to specify this.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+2.  The confidence on the proposal exceeds our threshold for
+    finalization.
+    
+3.  The number of `rounds` executed would be greater than
```
```suggestion
3.  The number of `rounds` executed is greater than
```

(It is a definitive termination condition.)

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+The algorithim only requires that nodes have computed the drift of
+observation of the passage of local time, not that that they have
+coordinated an absolute time with their peers.  For an implementation
+of a phase locked-loop feedback to measure local clock drift see [ntp][].
```
```suggestion
of a phase locked-loop feedback to measure local clock drift see [NTP](https://www.rfc-editor.org/rfc/rfc5905.html).
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+### Node receives information during round
+
+In the query step, the node is envisioned as packing information into
+the query to cut down on the communication overhead a query to each of
```
```suggestion
the query to cut down on the communication overhead of a query to each of
```

kaiserd replied:
(bump)

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+In the query step, the node is envisioned as packing information into
+the query to cut down on the communication overhead a query to each of
+this `k` nodes containing the node's own current opinion on the
```
```suggestion
these `k` nodes containing the node's own current opinion on the
```

kaiserd replied:
(bump)
Also, the information might be easier to parse if this sentence was split.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+the query to cut down on the communication overhead a query to each of
+this `k` nodes containing the node's own current opinion on the
+proposal ("YES", "NO", or "NONE").  The relation of the metadata
+is not currently analyzed to these opinions should be constrained.
```
Not sure what is meant here.

corpetty replied:
agreed, oddly worded sentence.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+The current algorithm doesn't describe how the initial opinions are formed.
+
+# Implementation status
+
```
Just as a note: Imo, this is section should be removed once the RFC advances to stable.
In a raw draft, it is helpful info.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+Python and Common Lisp are also in limited public review.
+
+
+# Interoperability
```
I'd call this section wire protocol, to make the purpose of this section explicit (see template).
Imo, it is fine, if the raw RFC does not specify the full wire protocol yet; but the future goal is for this section to specify the wire protocol.

kaiserd replied:
(The wire protocol section may contain a concise semantics description, as the one you list below. This is very helpful :))

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+the parity summations across network invariants often become easier to
+manipulate.
+
+# Sovereignty Considerations
```
I'd still call this `Security Considerations` to be consistent with other RFCs. 
Wdyt?

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+In practice, each honest node gossips its current opinion which
+reduces the number of messages that need to be gossiped for a given
+proposal.  The resulting impact on the privacy of the node's opinion
```
```suggestion
proposal. The resulting impact on the privacy of the node's opinion
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+## Security with respect to various Adversarial Models 
+
+Adversarial models have been tested for which the values for current
```
A reference to these tests would be good.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+Like a petulant child, an infantile adversary responds with the
+opposite vote of the honest majority on an opinion.
+
```
How does Glacier deal with this attacker?

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+consensus by being able to control `f` of `N` nodes with a out-of-band
+"supra-liminal" coordination mechansim.  Such adversaries use this
+coordinated behavior to delay or sway honest majority consensus.
+
```
What are the effects? Is Glacier safe in this model?

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+of the network state. Such an adversary can always chose to vote in
+the most efficient way to block the distributed consenus from
+finalizing.
+
```
Is Glacier safe in this model?

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+decisions to ones favorable to itself.  This adversary will, of
+course, choose to participate in an honest manner until defecting is
+most advantageous.
+
```
Is Glacier safe in this model?

kaiserd reviewed (commented):
I left some more comments :).
(Did not check the appendix.)

As @oskarth alread said, a comparison with other consensus mechansims would be helpful.
This could be added in a subsequent PR though.

Just as a reminder: RFCs should use [sembr](https://sembr.org/),
as pointed out by @D4nte .

corpetty commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+# Abstract
+
+This document specifies Glacier: a Byzantine tolerant binary decision
+agreement algorithm.  Glacier is a novel member of the Snow family
```
I'd suggest we add "in that it is a probabalistic leaderless BFT consensus algorithm that leverages mestability via network subsampling"

corpetty commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+This document specifies Glacier: a Byzantine tolerant binary decision
+agreement algorithm.  Glacier is a novel member of the Snow family
+that utilizes bounded memory for its execution.  We present an
+application context of the use of Glacier in a more efficient,
```
a more -> an

corpetty commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+## Leaderless Consensus Model Application to Shared State 
+
+Given an underlying binary consensus mechanism, one may quickly vote
+on the distributed, directed acyclic ledger graph of transactions.  If
```
no need to focus on what type of data we are coming to consensus on. An algorithm like this doesn't care.

corpetty commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+;; FIXME: find and justify an empirical confidence threshold 
+confidence_threshold
+  <-- 1  ;; BOGUS:  this should be a function of the network size
```
a confidence of 1 means that we have infinite evidence in a given direction, and thus, full confidence in a given decision.

AMozeika replied:
Suppose that at time $t=0$ \emph{all}  $N$ nodes in the network  are  in state Yes (or No), and there are no adversaries,  then  each node in the  \emph{Glacier} will stop updating after $\frac{\ell c}{K(1-c)}$ iterations, where $c\in(0,1)$ is the confidence  threshold, $\ell$ is the look-ahead parameter  and $K$ is  the query size. So $c$ has to be strictly less that $1$, otherwise number of iterations will diverge as $c$ approaches $1$.

corpetty commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+any Network State.  One of these components is the ability to come to
+consensus on a given proposal.  We need to replace resource intensive
+Nakomoto consensus mechanisms with those which belong to the class of
+more efficient ones rooted in leaderless decentralization.  By
```
I'd like a note on the fact that they are also less constrained than PoW and thus, more adaptive to a network's requirements.

corpetty commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+all information gathered during the query process retaining only the
+final opinion on the truth of the proposal.
+
+### Setup Parameters
```
I would definitely like a more formal definition of the variables involved, and a given greek identifier so we can write formulas and reference them more easily.

easye replied:
@corpetty I could use a little more instruction on how to revise: 

1.  A Greek identifier for each constant/parameter assigination?  LaTeX is still a little wonky, so I just did all the actual algorithm in straight Markdown.  I would use UTF-8 glyphs at the moment for this labeling.

2.  By "formal" do you mean some sort of type specification?  I eliminated the use of floats on purpose, so that currently all one needs is positive integer ratios.  We could put an upper limit on the size of the integers, but I don't have justification for this other than "what would reasonably fit" efficiently on contemporary architectures.

AMozeika replied:
On page 7 of  https://www.overleaf.com/read/vrqshkdxkfxz I have written mathematical description of the Glacier algorithm  and perhaps some of the notation introduced  can be also used in this document.

corpetty commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+where `w_i` is the weight of the `i`th peer.
+
+This weighted value would be used to reflect decisions mediating
```
maybe rephrase to something like:

"A given nodes weight can be calcualated in a multitude of ways and is left as an implementation detail or further modification. Such a weight could be derived from a multitude of values including:"

corpetty commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+2. [snow-ipfs](<https://ipfs.io/ipfs/QmUy4jh5mGNZvLkjies1RWM4YuvJh5o2FYopNPVYwrRVGV>)
+
+3. [snow*](<https://arxiv.org/abs/1906.08936v2>) Rocket Team, Maofan
```
it is reasonable to point to the Ava Labs whitepapers here, as they are the most up to date versions of the papers. https://www.avalabs.org/whitepapers

corpetty commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+# Appendix A: Alvaro's Exposition of Glacier
+
+In this appendix, we present Alvaro's original notion of Glacier,
```
I'm not sure we need to include this.

kaiserd replied:
Maybe remove this appendix and add a link in the introductory sections of this spec?

corpetty commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,695 @@
+
+the parity summations across network invariants often become easier to
+manipulate.
+
```
@kaiserd do you currently have a place for experimental results and analysis for a given specification? We have quite a bit of work here, but not sure how or where to add it to the specification. 

I'd previously thought it would be useful to add an additional level of specification after stable like "vetted" or "verified" or something that helps give confidence to the level of confirmation analysis for a given specification.

kaiserd replied:
@corpetty 

> @kaiserd do you currently have a place for experimental results and analysis for a given specification? We have quite a bit of work here, but not sure how or where to add it to the specification.

For now, I'd put such results in the appendix.
For future versions (draft, stable), results and analyses might be published in papers, which can be cited in the RFC.

> I'd previously thought it would be useful to add an additional level of specification after stable like "vetted" or "verified" or something that helps give confidence to the level of confirmation analysis for a given specification.

We could introduce a state.
Alternatively, we could introduce "verified" as a tag.

My suggestion would be to only promote RFCs to `stable` if they are already "verified";
either formally or at least by the interoperability of independent implementations.
This is what the IETF does. IETF, proposals can be in draft state for a long time.
TLS 1.3, for example, had 28 draft versions (over 4 years) before being promoted to a stable RFC.
Imo, this level of stability is expected by implementers when seeing a `stable` spec.

I'll open a discussion issue for this :)

corpetty reviewed (commented):
Overall the specification is accurately recorded. I think there remains some polish and re-structuring of the abstract and introduction. 

After that, I think it is suitable for a `raw` publish.

easye committed:
Reasonably complete edit addressing Corey/ksr's concerns

easye committed:
Non-functionally re-order the N3 semantics

easye committed:
Smooth syntax description

easye committed:
Take @kaiserd edit on "simple two-level"

AMozeika commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,714 @@
+    total_evidence 
+      <-- total_positive / total_votes
+    new_evidence 
+      <-- positive_votes / new_votes
```
Suppose that a node queried $K$ nodes but  all  $K$  haven't  responded.  In this case   positive_votes=0 and  new_votes=0 and we will get 0/0 division, i.e. new_evidence is undefined. Intuitively when we have no responses we have no evidence, but we need to decide on have to handle this situation.

danielSanchezQ replied:
In the implementation we just compute things if we at least 1 vote. https://github.com/logos-co/consensus-research/blob/101cf169e4665e6720ce6ee3c5bc582337b73cd1/consensus/claro/src/claro.rs#L248

oskarth commented:
@easye What's the state of this PR, why hasn't it been touched in 3 weeks and why isn't it merged yet?

corpetty commented:
I've picked this up now. I'll have a final review draft for a RAW publication done by EoW. It will be slimmed down in language and closer in format to other RFC specs.

corpetty commented:
@kaiserd @oskarth I've rewritten the RFC to a point that I think minor fixes are only required. We need to finish moving some of the implementations to new repos so they can be easily consumed. I'd appreciate an overview to see if I've missed something major w.r.t. formatting and information. 

If you don't mind additional edits after publication to Raw then we can just merge and update when repos are created and published.

corpetty commented:
also note if I've missed a point from previous critique then let me know. Treat me like I'm starting fresh on this thing despite it being open for such a long time.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+contributors:
+    - Álvaro Castro-Castilla 
+    - Mark Evenson
+    - Corey Petty
```
You could remove yourself as a contributor, now that you are the editor.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+3. After this sampling is finished, if there is a vote that has more than an `alpha` threshold, it accumulates one count for this opinion, as well as changes its opinion to this one. But, if a different opinion is received, the counter is reset to 1. If no threshold `alpha` is reached, the counter is reset to 0 instead.
+4. After several iterations of this algorithm, we will reach a threshold `beta`, and decide on that as final.
+
+Next, we will proceed to describe our new algorithm, based on Snowball. There are 3 main paths to achieving the initial target of improving the current state of the art:
```
When introducing ways of improving, I'd expect reasons why it is desirable to improve Snowball in the first place.

corpetty replied:
will include in a separate PR.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+- Find ways to extend the Snowball algorithm.
+- Find ways to modify the Snowball algorithm.
+- Find ways to strengthen the Snowball algorithm in particular scenarios, or as generic as possible.
+
```
Imo, these are a bit vague.
How is strengthening different from modifying?

corpetty replied:
deleted "modify" and just use strengthen for now

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+
+In our design, all 3 are attempted.
+
+We have identified a shortcoming of the Snowball algorithm that was a perfect starting point for devising improvements. The scenario is as follows:
```
I'd remove the text from 

> There are 3 main paths to achieving the initial target  [...]

up to here and just list:

> We have identified a shortcoming of the Snowball algorithm 

as a motivation.

Imo, having discovered a shortcoming and providing a solution is a strong motivation.
Starting with this as motivation also clearly says why it is desirable to do this.

I'd put the more general motivation into a scientific paper and explain

* why it is desirable/necessary to improve Snowflake
* why it makes sense to improve Snowflake and not design something new.

Without answers to these question, and the phrase

>  that was a perfect starting point for devising improvements

the motivation (to me) reads a bit like "searching for a problem in order to do be able to do research",
instead of solving an actual or expected problem that has impact in practice.
Imo, while the former is good to push the boundaries of theoretical research,
the latter is a stronger motivation for applied research resulting in a spec.

(Feel free to ignore this; just my thoughts while reading.)

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+fault tolerance.
+
+## Algorithmic concept
+Claro is an evolution of the Snowball BBA algorithm, in which we tackle specifically the perceived weakness described above. The main focus is going to be the counter and the triggering of the reset. Following, we elaborate the different modifications and features that have been added to the reference algorithm:
```
What does BBA stand for?

corpetty replied:
fixed

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+4. Additionally, we introduce a function for weighted sampling. This will allow the combination of different forms of weighting:
+    - Staking
+    - Heuristic reputation
+    - Manual reputation.
```
Links to these weighting methods would be helpful.

corpetty replied:
We don't have any, they're just options that could be used to inform weight.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+
+It’s worth delving a bit into the way the data is interpreted in order to reach a decision. Our approach is based conceptually on the paper [Confidence as Higher-Order Uncertainty](https://cis.temple.edu/~pwang/Publication/confidence.pdf), which describes a frequentist approach to decision certainty. The first-order certainty, measured by frequency, is caused by known positive evidence, and the h ...
+
+Intuitively, we are looking for a function of **`w`**, call it **`c`** for confidence, that satisfies the following conditions:
```
`w` has not been introduced yet.
Maybe just: "of evidence **`w`**"

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+`YES` or `NO`.  If it cannot form an opinion, it leaves its opinion as
+`NONE`.
+
+For now, we will ignore the proposal dessimination process and assume all nodes participating have an initial opinion to repond to within a given request. Further research will relax this assumption and analyze timing attacks on proposal propagation through the network. 
```
```suggestion
For now, we will ignore the proposal dissemination process and assume all nodes participating have an initial opinion to respond to within a given request. Further research will relax this assumption and analyze timing attacks on proposal propagation through the network. 
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+max_k_multiplier_power 
+  <-- 4
+    
+;;; Initial numbers of nodes queried in a round
```
```suggestion
;;; Initial number of nodes queried in a round
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+$$ 
+P(i) = \frac{w_i}{\sum_{j=0}^{j=N} w_j} 
+$$
+
```
Maybe it would help to mention again that $w$ is the evidence.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+
+Every time the threshold is not reached, we multiply *`k`* by a
+constant. In our experiments, we found that a constant of 2 works
+well, but what really matter is that it stays within that order of
```
```suggestion
well, but what really matters is that it stays within that order of
```

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+\text{Evidence per round}        & e_{round} \impliedby \frac{round\ positive\ votes}{round\ votes} \newline
+\end{array}
+$$
+
```
This LaTex block and the following algo have overlapping definitions (with different naming), e.g.
`e_{accum}`  and `total_evidence`
I'd just define these once.

kaiserd replied:
Reading further and tracking back, I see: The transition function below also matches the `total_evidence` and `new_evidence`.

Imo, specifying/introducing this as a transition function and two regimes makes this more complicated to understand.
I'd specify this using the following algorithm and then explain that this can be interpreted or seen as a transition function doing smooth transitions between two regimes.
I'd also explicitly mention that this means, these regimes are both active at the same time (kind of blend together),
with the transition making the influence of the one stronger and the other weaker. 
(If I understood this correctly.)

corpetty replied:
I'm going to leave this alone for right now, and improve in a future PR after a discussion with you and the team.

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+    alpha 
+      <-- doubt * ( 1 - confidence ) + certainty * confidence 
+    
+### Phase Three: Computation
```
Should this be "Transition Function", in line with the list above?

kaiserd commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+In order to eliminate the need for a step function (a conditional in
+the code), we introduce a transition function from one regime to the
+other. Our interest in removing the step function is twofold:
+
```
Imo, it would help to explain these regimes here. They were abstractly desribed in the background section, but it is not quite clear what they comprise exactly. If I understand correctly, they are governed by `alpha (accumulated confidence)` and `evidence`, respectively.

corpetty replied:
I agree, but will leave that to a future PR after some team discussion on how best to describe this transition.

kaiserd commented on content/docs/rfcs/38/README.md:
Imo, $\alpha_1$ and $\alpha_2$ are confusing names here, because they are chosen weight parameters and not derived from $\alpha$

corpetty replied:
agreed, will rethink them.

kaiserd commented on content/docs/rfcs/38/README.md:
```suggestion
A local round of Claro terminates in one of the following
```

kaiserd commented on content/docs/rfcs/38/README.md:
Is there any update?

kaiserd commented on content/docs/rfcs/38/README.md:
Imo, we don't need that section.

kaiserd commented on content/docs/rfcs/38/README.md:
I'd remove that section.

kaiserd reviewed (approved):
LGTM! :sparkle: Very nice. Background and algo spec are very clear now :).
I left some comments in-line.

Imo, this PR is ready to be merged as a raw RFC.
Follow up PRs on raw RFCs are no problem.
From [1/COSS](https://rfc.vac.dev/spec/1/):

> Changes to raw specifications can be unilateral and arbitrary. Those seeking to implement a raw specification should ask for it to be made a draft specification. Raw specifications have no contractual weight.

Nit: Vac RFCs should use [sembr](https://sembr.org/),
but we can do that in a follow-up PR.

danielSanchezQ commented on content/docs/rfcs/38/README.md:
```diff
@@ -0,0 +1,652 @@
+
+;;; maximum query rounds before termination
+max_rounds
+   <-- 997 ;; TODO justify
```
I'm not sure this is ok. In experimentations we considered a round **failed** if didn't reached consensus before round `100`.

corpetty replied:
added note, can change later with some justification.

danielSanchezQ commented on content/docs/rfcs/38/README.md:
should we just publish the `consensus-models` which is currently private?

corpetty replied:
you mentioned you wanted to publish each implementation into a separate repo. I'd say each implementation should be able to be used as a library to other people and not clumped together.

corpetty committed:
edit for review comments.

corpetty commented:
@kaiserd ready for merge based on response to previous commit's comments you and @danielSanchezQ made. 

Let's push to raw and then work on the larger more clear comments.

kaiserd commented:
LGTM :+1: I agree with pushing in the current state and addressing the rest in future PRs.

oskarth reviewed (approved):
LGTM as raw

Need to remove the random files that were edit then please go ahead and merge

Thanks for taking this over Corey!

kaiserd commented:
@corpetty The CI has been fixed. After removing the files @oskarth listed above and signing all commits, this PR should be mergable :).

kaiserd committed:
reset acid files to master

kaiserd committed:
remove .hugo_build.lock

kaiserd commented:
I rewrote the history to sign unsigned commits:

* remove merged-in commits form master (otherwise, we cannot cleanly rebase on master at the end)
* sigend all commits in this PR
* rebased on master

With this, we can close this PR.
Thanks to all who have contributed :).
